Pass CSRF token using Sec-WebSocket-Protocol header

iarenaza opened the pull request:
The proposed change tries to respect other values in the Sec-WebSocket-Protocol header, that the caller may have set in the `headers` option when calling `make-channel-socket-client!`

[Re: #418]

iarenaza commented:
I opted for prefixing the actual CSRF token value with `sente-csrf-token-`, to make it easier to find it and extract it on the server side. One of the reasons for that is because protocol names in the Sec-WebSockel-Protocol header cannot use a bunch of non-alphanumeric characters (like spaces, colons, dashes, etc). And we want the value to be easily recognizable, in case the client wants to send other protocol names in that same header (for some custom extra behavior).

ptaoussanis commented:
@iarenaza Thank you Iñaki, this looks good! Will try take a closer look + merge by end of tomorrow 👍

iarenaza commented:
Whoops. I had forgotten the other side of the connection: extracting the CRSF token from the Sec-WebSocket-Protocol header on the server side.

ptaoussanis commented:
@iarenaza Hi Iñaki, sorry for the delay on this.

Before merging, I'd like to step back a moment and re-evaluate what we're trying to do here. The current PR seems to be based on [this](https://github.com/taoensso/sente/issues/137#issuecomment-542281213) suggestion made earlier by @kaosko, right?

But it's actually not obvious to me what the motivation is/was behind that suggestion.

@kaosko Sorry for digging this up after so long, but could you maybe explain why you've previously suggested that the Sente CSRF token be sent in the `Sec-WebSocket-Protocol` header? (Context [here](https://github.com/taoensso/sente/issues/137#issuecomment-542281213)). It's not obvious to me what the benefit is, or that the [header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Sec-WebSocket-Protocol) is intended as a place to convey a CSRF token.

Even though the Protocol header seems to be an application-level header, it seems we'd be abusing the semantic intention of the header no? I imagine that stands a non-zero chance of causing issues in the future, and seems like it anyway could be surprising. What's the anticipated upside?

kaosko commented:
It's a JavaScript spec/API thing. It doesn't allow you to set any other headers for the WebSocket request object than "Sec-WebSocket-Protocol" so [if you want to pass auth/CSRF tokens in the headers, that's the way to do it](https://stackoverflow.com/questions/22383089/is-it-possible-to-use-bearer-authentication-for-websocket-upgrade-requests/35108078#35108078). I wouldn't call it abusing the semantic intention, perhaps merely stretching it, again as an _application_ level protocol. In the grand scheme of things, I guess not that different from manufacturer specific Wi-Fi Information Elements or DHCP Vendor Options.

Stepping back, [the anticipated upside according to OWASP (the most authoritative I could find) is that "A CSRF token must not be leaked in the server logs or in the URL"](https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#synchronizer-token-pattern).

Almost anything has a non-zero chance of causing issues in the future, but the server is certainly free to ignore any values in Sec-WebSocket-Protocol that it doesn't understand. Plenty of web pages, frameworks still sending various "X-*" headers while the naming practice was deprecated in 2012 and the servers may or may not act on those (and vice versa). That said, I'm not strongly arguing for adding this but don't see how it could hurt much either. Finally, I have used and still use the Sec-WebSocket-Protocol header for sending authentication tokens with the websocket request.

iarenaza commented:
@ptaoussanis Just to confirm that `1.21.0-alpha1` works as expected in our case :tada: . We no longer need to use the `:ws-custom` option on the client side :pray: 

Side comment: we still need to use some of the available "callback functions" (like `csrf-token-fn`) on the backend side (to deal with server side CSRF token retrieval and validation). But that is totally expected, because of the type of value that we are using as the CSRF token (as I already mentioned, we are using OpenID Connect ID tokens as the CSRF token)

So, thanks a lot @ptaoussanis !

ptaoussanis commented:
Great, thanks for the confirmation @iarenaza!

kaosko commented:
What about Aleph, Undertow etc? It seems that http-kit just returns [whatever you stick in :ring.websocket/protocol key of the ring response](https://github.com/http-kit/http-kit/blob/master/src/org/httpkit/server.clj), no?

ptaoussanis commented:
@iarenaza Hi Iñaki, thanks for the detailed+clear update 🙏

> P.S. We are so sorry that we didn't test it with Chrome before submitting this PR and wasted your valuable time 😞

No problem at all, I appreciate the time + effort you've put into this! I'm sorry that you've been running into issues with Sente and/or http-kit!

Re: how to proceed- I've unfortunately got limited knowledge + bandwidth on this topic since I'm currently quite deeply buried in a few other things.

I am planning a big batch of Sente work for June, but hopefully we can make progress here in the meantime!

Can we first back up a bit so that I can confirm my understanding?

@iarenaza Your motivation for #465 is described [here](https://github.com/taoensso/sente/pull/465#issuecomment-2782694714), right?

So:

- You already have a value that'd work as a CSRF token.
- But you don't want the client to send it to the server over a query param since you're concerned about the token leaking / getting logged.

So the goal is to get the CSRF token from the client to the server in a way that:

1. Doesn't leak the token.
2. Works reliably across different browsers and HTTP servers.
3. Minimizes the size/complexity of changes needed in Sente.

Is that right?

If we were to enumerate all the options we have for getting this done- what would that include?

1. Send the token in a `Sec-WebSocket-Protocol` header
2. Send the token in another header? Is that possible?
3. Send the token in the first WebSocket message? Is that possible?
4. Any other ideas?

I haven't had the opportunity to properly investigate this myself, so any input here would be very welcome!

Put briefly, what's currently the main impediment to (1)?

What do other/similar libraries tend to do?

iarenaza committed:
Pass CSRF token using Sec-WebSocket-Protocol header

The proposed change tries to respect other values in the
Sec-WebSocket-Protocol header, that the caller may have set in the
headers option when calling make-channel-socket-client!

Use the pre-ws-handshake option introduced in http-kit commit to add
the Sec-WebSocket-Protocol response header in the websocket
upgrade. Some browsers (like Google Chrome and Google Chrome derived
ones) fail to complete the upgrade if the upgrade response doesn't
include the Sec-WebSocket-Protocol with one of the sub-protocols
proposed in the request.

[Re: 465]

iarenaza commented:
Hi @ptaoussanis ,

I have finally extracted my original changes from my local test repos, cleaned them up and removed all the `prn`s, etc. And I have applied those changes on top of the sente and http-kit main/master branches (as of today, 2025.11.26). I have kept the original sente and http-kit versions in their respective `project.clj` files, to be able to test them more easily using the Leiningen "checkouts" feature.

Then I have created a separate git repo, using the `example-project` code from the sente project, where I have cloned my forked repositories in the `checkouts` directory.

This is what I have done:

1. Clone the https://github.com/iarenaza/sente-example-project.git repository locally.

2. Create a `checkouts` directory inside the cloned repository.

3. Inside the `checkouts` directory, clone these two other repositories, that contain my patches:

   - https://github.com/iarenaza/sente.git
   - https://github.com/iarenaza/http-kit.git

4. After the previous steps, we should have the following directory structure:

   ```
   .
   ├── checkouts
   │   ├── http-kit
   │   └── sente
   ├── resources
   │   └── public
   └── src
       └── example
   ```

5. Now we can run `lein start` or `lein start-dev`, as in the original `example-project` from the sente project, from the top of the `sente-example-project` cloned repository.

6. I tested it with both Firefox 144.0 and Chromium "142.0.7444.175 (Official Build) built on Debian GNU/Linux 13 (trixie) (64-bit)", both running on Debian Trixie, and it worked as expected.

Do the changes look like something that you could consider including?

ptaoussanis commented:
@iarenaza Hi Iñaki,

> I haven't spent time on that, because I wanted to see if you were open to this approach at all, before committing to a non-trivial amount of research :-)

Gotcha, thanks for clarifying- and for all your work on this 👍

So in its current form, it's difficult to judge. I'm not opposed to the idea in principle, but unfortunately would need the non-trivial details in place to be able to properly evaluate.

The short / high-level version is this:

- I'm open to sensible non-breaking additions/improvements to http-kit and Sente.
- In both cases also _potentially_ open to breaking changes iff there's a strong  security/correctness need, and a very strong argument for why non-breaking alternatives wouldn't be sufficient.

So the key questions I'd want addressed before I'd be able to say if the path is viable or not:

* Is there a way we could do this without breaking anyone?
* Is there a way we could do this while also offering support to other servers?

If it's very costly to determine the above definitively, then it'd at least be helpful as a starting point to estimate.

Hope that's some help at least!
